### PR TITLE
29b multiclass brca classifier

### PR DESCRIPTION
**Pull request by teddyhla**
hi guys, 

this is a bit basic and a bit hacky: 

However, I have done a PR 

1. ingest the labelled brca data as json
2. currently uses pandas to do some cleaning [will fix in the future]
3. then tokenize using bert
4. train as multi class. classifier
5. evalute as multi class classifier with relevant metrics
6. the metrics are saved into mlflow. 

thanks.

### CODE REVIEW

**Review by drjzhn — COMMENTED** [review]
Nice work Teddy.

Agree with your todos, but let's move to refactoring to the base package. I'll give this a go based on your (and Sophie's) scripts, and let's catch up when you're back.

File: projects/test_multiclass_classifier/multiclassifier.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+DATA_PATH = "data/breast_brca_labelled.json"
+FRAC = 1
+RANDOM_STATE = 42
+TRAIN_TEST_SPLITPERC = 0.8
+CHOSEN_MODEL = "bert-base-uncased"
+BATCH_SIZE = 8
+SAVELOCATION = "mclassout"
+
+
+# sense check NEED TO GO TO MLFLOW
```
Comment (sophie22) [inline projects/test_multiclass_classifier/multiclassifier.py:39]: could do something like 
```
class_counts = df["ClassLabel"].value_counts()
    for class_label, count in class_counts.items():
        mlflow.log_param(f"class_{class_label}_count", count)
```

File: projects/test_multiclass_classifier/multiclassifier.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    id2label = labelsdict
+    label2id = {value: idx for idx, value in id2label.items()}
+    df["category"] = df["label"].apply(lambda x: label2id[x])
+    df["category"] = df["category"].astype("int")
+    df.columns = ["text", "category", "labels"]
+    print(f"preprocessing complete")
+    return df, labels, num_labels, id2label, label2id
+
+
+def create_hfds(data, frac, random_state, splitfrac):
```
[the remark this answers is not in the document]
Comment (sophie22) [reply projects/test_multiclass_classifier/multiclassifier.py:65]: HuggingFace dataset?

File: projects/test_multiclass_classifier/multiclassifier.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        random_state (_type_, optional): _description_. Defaults to RANDOM_STATE.
+        splitfrac (_type_, optional): _description_. Defaults to TRAIN_TEST_SPLITPERC.
+
+    Returns:
+        _type_: _description_
+    """
+    print(
+        f"creating training and testing data by shuffling the {frac} with a random state of {random_state}"
+    )
+    data_shuffled = data.sample(frac=frac, random_state=random_state).reset_index(
```
Comment (sophie22) [inline projects/test_multiclass_classifier/multiclassifier.py:80]: what is the benefit of shuffling?

File: projects/test_multiclass_classifier/multiclassifier.py [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        drop=True
+    )
+    split_index = int(splitfrac * len(data_shuffled))
+    train_df = data_shuffled.iloc[:split_index]
+    test_df = data_shuffled.iloc[split_index:]
+
+    ds = DatasetDict()
+    train = Dataset.from_pandas(train_df)
+    val = Dataset.from_pandas(test_df)
```
Comment (sophie22) [inline projects/test_multiclass_classifier/multiclassifier.py:89]: in future, do we need 3 sets? train and val for each epoch and a test set reserved for calculating performance metrics at the end?

File: projects/test_multiclass_classifier/multiclassifier.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    macro_precision, macro_recall, macro_f1, _ = precision_recall_fscore_support(
+        labels, predictions, average="macro", zero_division=0
+    )
+    conf_matrix = confusion_matrix(labels, predictions)
+
+    conf_matrix_dict = {
+        f"cm_{i}_{j}": conf_matrix[i][j]
+        for i in range(conf_matrix.shape[0])
+        for j in range(conf_matrix.shape[1])
+    }
```
[the remark this answers is not in the document]
Comment (sophie22) [reply projects/test_multiclass_classifier/multiclassifier.py:131]: in future we could log these to mlflow as artefacts (i haven't figured out if it's here or when trainer.evaluate() is called)

File: projects/test_multiclass_classifier/multiclassifier.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+def train_model(
+    model,
+    training_dataset,
+    evaluation_dataset,
+    metrics=compute_metrics,
+):
+
+    trainingargs = TrainingArguments(
+        output_dir="mclassout",
```
Comment (sophie22) [inline projects/test_multiclass_classifier/multiclassifier.py:153]: after refactoring, some of these could go into a config file

File: projects/test_multiclass_classifier/multiclassifier.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+        tokenizer, model = prepare_model_and_tokeniser(
+            model_name=CHOSEN_MODEL,
+            numberoflabels=num_labels,
+            l2i=label2id,
+            i2l=id2label,
+        )
+        ds_enc = ds.map(
+            tokenize_encode,
+            fn_kwargs={"tokenizer": tokenizer},  # NOTE this!
```
Comment (sophie22) [inline projects/test_multiclass_classifier/multiclassifier.py:201]: it's so cool that you can do so many dataset transformations in one command.
why is the highlighted line necessary? - is that to pass the input to the tokenize_encode function?

would it be easier to do tokenisation first and remove columns next?
```
train_tokenized = train_dataset.map(
        lambda x: tokenize_function(x, tokenizer),
        batched=True,
    )
```

File: projects/test_multiclass_classifier/multiclassifier.py [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    mlflow.set_experiment(experiment_name)
+    print("ML Flow experiement set up ")
+
+    df, labels, num_labels, id2label, label2id = load_process_pretokeniser(
+        path=DATA_PATH
+    )
+    print("Data is read, labels dict generated.")
+
+    ds = create_hfds(
+        data=df, frac=FRAC, random_state=RANDOM_STATE, splitfrac=TRAIN_TEST_SPLITPERC
```
Comment (sophie22) [inline projects/test_multiclass_classifier/multiclassifier.py:187]: I wonder if splitting the dataset could be the last step after data preparation and tokenisation.

Is there performance difference if we split first then tokenise versus tokenise all then split?

File: projects/test_multiclass_classifier/multiclassifier.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        mlflow.transformers.log_model(
+            transformers_model={
+                "model": model,
+                "tokenizer": tokenizer,
+            },
+            artifact_path="bert_model",
+            task="text-classification",  # NOTE:DO not change this!
+            signature=infer_signature(sample_input, torch.tensor([[0.1, 0.9]])),
+        )
+        trainer.model = trainer.model.cpu()
```
Comment (sophie22) [inline projects/test_multiclass_classifier/multiclassifier.py:226]: why is this necessary? I saw it in a lot of tutorials as well, but my scripts uses MPS without explicitly being told

File: projects/test_multiclass_classifier/multiclassifier.py [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+
+    trainingargs = TrainingArguments(
+        output_dir="mclassout",
+        evaluation_strategy="epoch",
+        learning_rate=2e-5,
+        per_device_train_batch_size=8,
+        per_device_eval_batch_size=8,
+        num_train_epochs=3,
+        weight_decay=0.01,
+        use_cpu=True,
```
Comment (sophie22) [inline projects/test_multiclass_classifier/multiclassifier.py:160]: if you remove this where is your model training running?

File: projects/test_multiclass_classifier/requirements.txt
[the remark this answers is not in the document]
Comment (teddyhla) [reply projects/test_multiclass_classifier/requirements.txt:1]: Thanks Jo, sorry about this. what is your recommendation to do this? [pipreqs?](https://github.com/bndr/pipreqs)

File: projects/test_multiclass_classifier/multiclassifier.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (teddyhla) [reply projects/test_multiclass_classifier/multiclassifier.py:65]: hugging face datasets. i have to add descriptions and docstrings for the functions in this pr.

File: projects/test_multiclass_classifier/multiclassifier.py (hunk 3 above)
Comment (teddyhla) [reply projects/test_multiclass_classifier/multiclassifier.py:80]: it is probably unnecessary. But sometimes especially in multi-class datasets, the data may be ordered based on class labels. The shuffling take care of that. If the data instances are already random and not orgnaised in certain way, it doesnt hurt to reshuffle again.  can take this out if considered verbose.

File: projects/test_multiclass_classifier/multiclassifier.py (hunk 4 above)
Comment (teddyhla) [reply projects/test_multiclass_classifier/multiclassifier.py:89]: That's true. In classical deep learning, my understanding is that train is to train params, and val is for hyperparam tuning and test for metrics. i am not sure how hugging face uses val set for. trawling through hf forum, it seems to suggest soemthing along the lines of validation at end of each epoch. [see this discussion on hf forum](https://discuss.huggingface.co/t/validation-vs-test-with-transformers-trainer/17961/3)

File: projects/test_multiclass_classifier/multiclassifier.py (hunk 8 above)
Comment (teddyhla) [reply projects/test_multiclass_classifier/multiclassifier.py:187]: good question. from classical machine learning (aka structured data), i pick up a habit (i think other practitioners do too) to split train and test early. therefore whatever transformations you apply dont cause a data leakage (e.g. ,divding by mean).  i am not sure this prudence is needed for text.

File: projects/test_multiclass_classifier/multiclassifier.py (hunk 9 above)
Comment (teddyhla) [reply projects/test_multiclass_classifier/multiclassifier.py:226]: i cannot explain why.  was stuck in a bug and only when trainer.model.cpu is done that i got out of it. jo make a suggestion abotu declaring device earlier. I think that is probably much cleaner. need to however link `device` with model.cpu() however

File: projects/test_multiclass_classifier/multiclassifier.py (hunk 10 above)
Comment (teddyhla) [reply projects/test_multiclass_classifier/multiclassifier.py:160]: good questin. i will have a look

File: projects/test_multiclass_classifier/requirements.txt
[the remark this answers is not in the document]
Comment (sophie22) [reply projects/test_multiclass_classifier/requirements.txt:1]: I tend to add packages manually to the requirements file as I need to install them for each project, that way the list is minimised and doesn't include dependencies of the required package (which will be installed anyway).

File: projects/test_multiclass_classifier/multiclassifier.py (hunk 3 above)
Comment (sophie22) [reply projects/test_multiclass_classifier/multiclassifier.py:80]: I like it, thanks for the explanation!
